skimage/morphology/extrema.py

```diff
@@ -0,0 +1,349 @@
+"""extrema.py - local minima and maxima
+
+This module provides functions to find local maxima and minima of an image.
+Here, local maxima (minima) are defined as connected sets of pixels with equal
+grey level which is strictly greater (smaller) than the grey level of all
+pixels in direct neighborhood of the connected set. In addition, the module
+provides the closely related functions h-maxima and h-minima.
+
+Soille, P. (2003). Morphological Image Analysis: Principles and Applications
+(2nd ed.). Springer-Verlag New York, Inc.
+
+Original author: Thomas Walter
+"""
+
+import greyreconstruct
+import numpy as np
+
+
+def _add_constant_clip(img, const_value):
+    """Adds a constant to the image and handles overflow issues
+    for integer typed images.
+    """
+    array_info = np.iinfo(img.dtype)
+    max_dtype = array_info.max
+    min_dtype = array_info.min
+
+    if const_value > (max_dtype - min_dtype):
+        raise ValueError("The added constant is not compatible"
+                         "with the image data type.")
+
+    result = img + const_value
+    result[img > max_dtype-const_value] = max_dtype
+    return(result)
+
+
+def _subtract_constant_clip(img, const_value):
+    """Subtracts a constant from the image and handles overflow issues
+    for integer typed images.
+    """
+    array_info = np.iinfo(img.dtype)
+    max_dtype = array_info.max
+    min_dtype = array_info.min
+    if const_value > (max_dtype-min_dtype):
+        raise ValueError("The subtracted constant is not compatible"
+                         "with the image data type.")
+
+    result = img - const_value
+    result[img < (const_value + min_dtype)] = min_dtype
+    return(result)
+
+
+def h_maxima(img, h, selem=None):
+    """Determines all maxima of the image with height >= h.
+
+    The local maxima are defined as connected sets of pixels with equal
+    grey level strictly greater than the grey level of all pixels in direct
+    neighborhood of the set.
+
+    A local maximum M of height h is a local maximum for which
+    there is at least one path joining M with a higher maximum on which the
+    minimal value is f(M) - h (i.e. the values along the path are not
+    decreasing by more than h with respect to the maximum's value) and no
+    path for which the minimal value is greater.
+
+    Parameters
+    ----------
+    img : ndarray
+        The input image for which the maxima are to be calculated.
+    h : unsigned integer
+        The minimal height of all extracted maxima.
+    selem : ndarray
+        The neighborhood expressed as a 2-D array of 1's and 0's.
```

Comment:
Probably. `selem` is passed to `greyreconstruct` (so I do not access this myself), I therefore took the description from `greyreconstruct`. I did not see a nD test, but as far as I can tell from the code, it should work in n dimensions. I will make a 3D test to see whether this truly works and we can then change the two descriptions.